### PR TITLE
[User Schedule] Add cacheBefore, cacheAfter, and setMemoryType

### PR DESCRIPTION
**Pull request by rdspring1**
This PR adds the `cache_before`, `cache_after`, and `set_memory_type` functions to user scheduling.

Note: `LoadStoreOpType::CpAsyncBulkTensorTile` is changed to `LoadStoreOpType.tma` for easier readability.

### PR List:
1. #2282 
2. #2283 
3. #2284 **<<<  YOU ARE HERE**
4. #2285
5. #2286
6. #2287
7. #2288

### CODE REVIEW

File: csrc/python_frontend/python_bindings.cpp [hunk 1]
```diff
@@ -421,6 +421,27 @@ void initNvFuserPythonBindings(PyObject* module) {
...
+  py::enum_<LoadStoreOpType>(nvfuser, "LoadStoreOpType")
+      .value("set", LoadStoreOpType::Set)
+      .value("load_matrix", LoadStoreOpType::LdMatrix)
+      .value("load_matrix_transpose", LoadStoreOpType::LdMatrixTranspose)
+      .value("cp_async", LoadStoreOpType::CpAsync)
+      .value("tma", LoadStoreOpType::CpAsyncBulkTensorTile);
+
+  //! CacheOp used for scheduling
+  py::enum_<CacheOp>(nvfuser, "CacheOp")
+      .value("none", CacheOp::Unspecified)
```
Comment (jjsjann123) [inline csrc/python_frontend/python_bindings.cpp:434]: nitpick, should we stick with `unspecified` instead?

File: csrc/python_frontend/python_bindings.cpp [hunk 2]
```diff
@@ -2921,6 +2942,65 @@ void initNvFuserPythonBindings(PyObject* module) {
...
+            "Attempting to use a SchedOperators Op prior to definition!");
+        FusionDefinition* fd = self.fusion_definition;
+        TensorView* input_tv =
+            fd->getFusionState(tensor.index)->template as<TensorView>();
+        TensorView* output_tv = input_tv->cacheAfter(op_type, cache_op);
+        Tensor output = fd->defineTensor(tensor.dims);
+        NVF_CHECK(
+            output.index == fd->numFusionStates(),
+            "Fusion State index does not match the size!");
+        fd->addFusionState(output_tv);
```
Comment (jjsjann123) [inline csrc/python_frontend/python_bindings.cpp:2962]: this pattern is showing up enough times I think we can add a function in FusionDefinition as `Tensor FusionDefinition::addTensor(TensorView* tv)`.

I think `tensor.dims == output_tv->nDims()`.

File: csrc/python_frontend/python_bindings.cpp [hunk 3]
```diff
@@ -2921,6 +2942,65 @@ void initNvFuserPythonBindings(PyObject* module) {
...
+            "Attempting to use a SchedOperators Op prior to definition!");
+        FusionDefinition* fd = self.fusion_definition;
+        TensorView* input_tv =
+            fd->getFusionState(tensor.index)->template as<TensorView>();
+        TensorView* output_tv = input_tv->cacheBefore(op_type);
+        Tensor output = fd->defineTensor(tensor.dims);
+        NVF_CHECK(
+            output.index == fd->numFusionStates(),
+            "Fusion State index does not match the size!");
+        fd->addFusionState(output_tv);
```
Comment (jjsjann123) [inline csrc/python_frontend/python_bindings.cpp:2984]: ditto.

Also there's another use in reduction_factor

File: csrc/python_frontend/python_bindings.cpp (hunk 2 above)
Comment (jjsjann123) [reply csrc/python_frontend/python_bindings.cpp:2962]: Giving that this is the only use case for addFusionState, maybe we can change that one instead.

**Review by jjsjann123 — APPROVED** [review]
Since this PR is only adding new things, we don't need to bump version.txt. LGTM with some minor comments

File: csrc/python_frontend/python_bindings.cpp (hunk 1 above)
Comment (rdspring1) [reply csrc/python_frontend/python_bindings.cpp:434]: 🤷🏼‍♂️ Lol. I can't consistently spell `unspecified` without an IDE.

**Comment by rdspring1** [comment]
!build

**Review by jjsjann123 — APPROVED** [review]

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Pull request merged**